registry/api/v2/descriptors.go

```diff
@@ -637,17 +588,9 @@ var routeDescriptors = []RouteDescriptor{
 									ErrorCodeBlobUnknown,
 								},
 							},
-							{
-								StatusCode:  http.StatusUnauthorized,
-								Description: "The client does not have permission to push to the repository.",
-								Body: BodyDescriptor{
-									ContentType: "application/json; charset=utf-8",
-									Format:      errorsBody,
-								},
-								ErrorCodes: []errcode.ErrorCode{
-									errcode.ErrorCodeUnauthorized,
-								},
-							},
+							unauthorizedResponseDescriptor,
+							repositoryNotFoundResponseDescriptor,
```

Comment:
It's not about whether the registry has any blobs/manifests in a repository, it's about whether the access controller has any record of the repository. The access controllers do their own existence checks on these records. If they always return 403s, this may confuse the client which may think that the resource exists but they are not authorized to access it. A 404 signals to the client that the resource does not exist and so *of course* they can't perform any action on it. Some access controllers may choose to return 404s when the client does not have minimal access (e.g., at least pull access) to an existing resource to say "this resource does not exist (as far as you are concerned)".

I'm not saying every access controller has to work this way, but some may want to. For your push example, for instance, if you (authenticated as `dmcgowan`) are trying to push to a repository named `dmcgowan/data-analyzer` but it does not exist yet on Docker Hub, the Docker Hub authorizer (currently) creates the record for you because you are the owner. However imagine if Docker Hub required you to create/initialize the repository on the website first and you tried this. If the registry returned you a 403 it might lead to confusion: "why would I not be allowed to push? it has my name in it!?". But if it returned a 404 you would realize that it does not exist yet and you should create it.